File: free-todo-frontend/scripts/check_code_lines.js
#!/usr/bin/env node
/**
 * Check effective TypeScript/TSX code lines (excluding blank lines and comments).
 * Files over the limit are reported and the script exits non-zero.
 *
 * Usage:
 *   # Scan the entire directory (standalone)
 *   node check_code_lines.js [--include dirs] [--exclude dirs] [--max lines]
 *
 *   # Check specified files (pre-commit mode)
 *   node check_code_lines.js [options] file1.ts file2.tsx ...
 *
 * Examples:
 *   # Scan the whole frontend directory
 *   node check_code_lines.js --include apps,components,lib --exclude lib/generated --max 500
 *
 *   # Check specific files (pre-commit passes staged files)
 *   node check_code_lines.js apps/chat/ChatPanel.tsx apps/todo/TodoList.tsx
 */

const { existsSync, readdirSync, readFileSync } = require("node:fs");
const { dirname, isAbsolute, join, relative, resolve } = require("node:path");

// Script directory (CommonJS)
// In CommonJS, __dirname and __filename are available by default.

// Default configuration
const DEFAULT_INCLUDE = ["apps", "components", "electron", "lib"];
const DEFAULT_EXCLUDE = ["lib/generated"];
const DEFAULT_MAX_LINES = 500;

/**
 * @typedef {Object} Config
 * @property {string[]} includeDirs
 * @property {string[]} excludeDirs
 * @property {number} maxLines
 * @property {string[]} files - Explicit file list
 */

/**
 * Parse CLI arguments.
 * @returns {Config}
 */
function parseArgs() {
  const args = process.argv.slice(2);
  let includeDirs = DEFAULT_INCLUDE;
  let excludeDirs = DEFAULT_EXCLUDE;
  let maxLines = DEFAULT_MAX_LINES;
  /** @type {string[]} */
  const files = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--include" && args[i + 1]) {
      includeDirs = args[i + 1]
        .split(",")
        .map((d) => d.trim())
        .filter(Boolean);
      i++;
    } else if (arg === "--exclude" && args[i + 1]) {
      excludeDirs = args[i + 1]
        .split(",")
        .map((d) => d.trim())
        .filter(Boolean);
      i++;
    } else if (arg === "--max" && args[i + 1]) {
      maxLines = parseInt(args[i + 1], 10);
      i++;
    } else if (!arg.startsWith("--")) {
      // Positional args are treated as file paths
      files.push(arg);
    }
  }

  return { includeDirs, excludeDirs, maxLines, files };
}

/**
 * Check whether a line is a comment-only line.
 *
 * Rule: after trim(), lines starting with //, /*, *, or * / are comments.
 * @param {string} line
 * @returns {boolean}
 */
function isCommentLine(line) {
  const trimmed = line.trim();
  return (
    trimmed.startsWith("//") ||
    trimmed.startsWith("/*") ||
    trimmed.startsWith("*") ||
    trimmed.startsWith("*/")
  );
}

/**
 * Count effective code lines (excluding blank lines and comments).
 * @param {string} filePath
 * @returns {number}
 */
function countCodeLines(filePath) {
  try {
    const content = readFileSync(filePath, "utf-8");
    const lines = content.split("\n");
    let codeLines = 0;

    for (const line of lines) {
      const trimmed = line.trim();
      // Skip blank lines
      if (!trimmed) {
        continue;
      }
      // Skip comment-only lines
      if (isCommentLine(line)) {
        continue;
      }
      // Counted line
      codeLines++;
    }

    return codeLines;
  } catch (error) {
    console.error(`Warning: failed to read file ${filePath}: ${error}`);
    return 0;
  }
}

/**
 * Normalize path separators to / (Windows-friendly).
 * @param {string} p
 * @returns {string}
 */
function normalizePath(p) {
  return p.replace(/\\/g, "/");
}

/**
 * Determine whether a file should be checked.
 * @param {string} relPath
 * @param {string[]} includeDirs
 * @param {string[]} excludeDirs
 * @returns {boolean}
 */
function shouldCheckFile(relPath, includeDirs, excludeDirs) {
  // Normalize to / separators
  const normalizedPath = normalizePath(relPath);

  // Check include directories
  const inInclude = includeDirs.some((inc) =>
    normalizedPath.startsWith(normalizePath(inc))
  );
  if (!inInclude) {
    return false;
  }

  // Check exclude directories
  const inExclude = excludeDirs.some((exc) =>
    normalizedPath.startsWith(normalizePath(exc))
  );
  if (inExclude) {
    return false;
  }

  return true;
}

/**
 * Recursively walk a directory for .ts and .tsx files.
 * @param {string} dir
 * @returns {Generator<string>}
 */
function* walkDir(dir) {
  try {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        // Skip node_modules and hidden directories
        if (entry.name === "node_modules" || entry.name.startsWith(".")) {
          continue;
        }
        yield* walkDir(fullPath);
      } else if (entry.isFile()) {
        if (entry.name.endsWith(".ts") || entry.name.endsWith(".tsx")) {
          yield fullPath;
        }
      }
    }
  } catch {
    // Ignore inaccessible directories
  }
}

/**
 * Get the list of files to check.
 * @param {Config} config
 * @param {string} rootDir
 * @returns {string[]}
 */
function getFilesToCheck(config, rootDir) {
  /** @type {string[]} */
  const filesToCheck = [];

  if (config.files.length > 0) {
    // Mode 1: Check specified files (pre-commit mode)
    for (const fileStr of config.files) {
      // Handle relative and absolute paths
      const filePath = isAbsolute(fileStr) ? fileStr : resolve(fileStr);

      // Only check .ts/.tsx files
      if (!filePath.endsWith(".ts") && !filePath.endsWith(".tsx")) {
        continue;
      }

      // Skip missing files
      if (!existsSync(filePath)) {
        continue;
      }

      // Get path relative to frontend root
      /** @type {string} */
      let relPath;
      try {
        relPath = relative(rootDir, filePath);
        // If rel path starts with "..", it's outside the frontend directory
        if (relPath.startsWith("..")) {
          continue;
        }
      } catch {
        continue;
      }

      // Check include/exclude filters
      if (shouldCheckFile(relPath, config.includeDirs, config.excludeDirs)) {
        filesToCheck.push(filePath);
      }
    }
  } else {
    // Mode 2: Scan entire directory (standalone mode)
    for (const filePath of walkDir(rootDir)) {
      const relPath = relative(rootDir, filePath);
      if (shouldCheckFile(relPath, config.includeDirs, config.excludeDirs)) {
        filesToCheck.push(filePath);
      }
    }
  }

  return filesToCheck;
}

/**
 * Main entrypoint.
 * @returns {number}
 */
function main() {
  const config = parseArgs();

  // Frontend root (script lives in free-todo-frontend/scripts/)
  const rootDir = dirname(__dirname);

  // Collect files to check
  const filesToCheck = getFilesToCheck(config, rootDir);

  if (filesToCheck.length === 0) {
    if (config.files.length > 0) {
      // No matching files in pre-commit mode
      return 0;
    }
    return 0;
  }

  // Collect violations
  /** @type {Array<{ path: string; lines: number }>} */
  const violations = [];

  for (const filePath of filesToCheck) {
    const relPath = relative(rootDir, filePath);
    const codeLines = countCodeLines(filePath);
    if (codeLines > config.maxLines) {
      violations.push({ path: relPath, lines: codeLines });
    }
  }

  // Output results
  if (violations.length > 0) {
    console.log(
      `[ERROR] The following files exceed ${config.maxLines} code lines:`
    );
    violations.sort((a, b) => a.path.localeCompare(b.path));
    for (const { path, lines } of violations) {
      console.log(`  ${path} -> ${lines} lines`);
    }
    return 1;
  }
  return 0;
}

process.exit(main());

File: lifetrace/scripts/check_code_lines.py
#!/usr/bin/env python3
"""
Check effective Python code lines (excluding blank lines and comments).
Files over the limit are reported and the script exits non-zero.

Usage:
    # Scan the whole directory (standalone)
    python check_code_lines.py [--include dirs] [--exclude dirs] [--max lines]

    # Check specific files (pre-commit mode)
    python check_code_lines.py [options] file1.py file2.py ...

Examples:
    # Scan the entire lifetrace directory
    python check_code_lines.py --include lifetrace --exclude lifetrace/__pycache__,lifetrace/dist --max 500

    # Check specific files (pre-commit passes staged files)
    python check_code_lines.py lifetrace/routers/chat.py lifetrace/services/todo.py
"""

import argparse
import sys
from pathlib import Path

# Default configuration
DEFAULT_INCLUDE = ["lifetrace"]
DEFAULT_EXCLUDE = [
    "lifetrace/__pycache__",
    "lifetrace/dist",
    "lifetrace/migrations/versions",
]
DEFAULT_MAX_LINES = 500


def parse_args() -> argparse.Namespace:
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(
        description="Check effective Python code lines (excluding blank lines and comments)."
    )
    parser.add_argument(
        "files",
        nargs="*",
        help="Files to check (if omitted, scan the entire directory).",
    )
    parser.add_argument(
        "--include",
        type=str,
        default=",".join(DEFAULT_INCLUDE),
        help=(
            f"Comma-separated directory prefixes to include (default: {','.join(DEFAULT_INCLUDE)})"
        ),
    )
    parser.add_argument(
        "--exclude",
        type=str,
        default=",".join(DEFAULT_EXCLUDE),
        help=(
            f"Comma-separated directory prefixes to exclude (default: {','.join(DEFAULT_EXCLUDE)})"
        ),
    )
    parser.add_argument(
        "--max",
        type=int,
        default=DEFAULT_MAX_LINES,
        help=f"Maximum allowed code lines (default: {DEFAULT_MAX_LINES}).",
    )
    return parser.parse_args()


def count_code_lines(file_path: Path) -> int:
    """
    Count effective code lines (excluding blank lines and comment-only lines).

    Rules:
    - Blank lines (strip() == ""): not counted
    - Lines starting with "#": not counted
    - All other lines: counted
    """
    code_lines = 0
    try:
        with open(file_path, encoding="utf-8") as f:
            for line in f:
                stripped = line.strip()
                # Skip blank lines
                if not stripped:
                    continue
                # Skip comment-only lines
                if stripped.startswith("#"):
                    continue
                # Counted line
                code_lines += 1
    except (OSError, UnicodeDecodeError) as e:
        print(f"Warning: failed to read file {file_path}: {e}", file=sys.stderr)
        return 0
    return code_lines


def should_check_file(
    file_path: Path, root_dir: Path, include_dirs: list[str], exclude_dirs: list[str]
) -> bool:
    """
    Determine whether a file should be checked.

    Args:
        file_path: File path
        root_dir: Project root directory
        include_dirs: Directory prefixes to include
        exclude_dirs: Directory prefixes to exclude

    Returns:
        True if the file should be checked; otherwise False
    """
    # Get path relative to the project root
    try:
        rel_path = file_path.relative_to(root_dir)
    except ValueError:
        return False

    # Normalize to forward slashes to avoid Windows separator issues
    rel_path_str = str(rel_path).replace("\\", "/")

    # Check include directories
    in_include = any(rel_path_str.startswith(inc.replace("\\", "/")) for inc in include_dirs)
    if not in_include:
        return False

    # Check exclude directories
    in_exclude = any(rel_path_str.startswith(exc.replace("\\", "/")) for exc in exclude_dirs)
    return not in_exclude


def get_files_to_check(
    args: argparse.Namespace, root_dir: Path, include_dirs: list[str], exclude_dirs: list[str]
) -> list[Path]:
    """
    Get the list of files to check.

    Args:
        args: Parsed command-line arguments
        root_dir: Project root directory
        include_dirs: Directory prefixes to include
        exclude_dirs: Directory prefixes to exclude

    Returns:
        List of file paths to check
    """
    files_to_check: list[Path] = []

    if args.files:
        # Mode 1: Check specified files (pre-commit mode)
        for file_str in args.files:
            file_path = Path(file_str).resolve()
            # Only check .py files
            if file_path.suffix != ".py":
                continue
            # Skip missing files
            if not file_path.exists():
                continue
            # Check include/exclude filters
            if should_check_file(file_path, root_dir, include_dirs, exclude_dirs):
                files_to_check.append(file_path)
    else:
        # Mode 2: Scan entire directory (standalone mode)
        for py_file in root_dir.rglob("*.py"):
            if should_check_file(py_file, root_dir, include_dirs, exclude_dirs):
                files_to_check.append(py_file)

    return files_to_check


def main() -> int:
    """Main entrypoint."""
    args = parse_args()

    # Parse arguments
    include_dirs = [d.strip() for d in args.include.split(",") if d.strip()]
    exclude_dirs = [d.strip() for d in args.exclude.split(",") if d.strip()]
    max_lines = args.max

    # Project root (script lives in lifetrace/scripts/)
    script_dir = Path(__file__).resolve().parent
    root_dir = script_dir.parent.parent

    # Collect files to check
    files_to_check = get_files_to_check(args, root_dir, include_dirs, exclude_dirs)

    if not files_to_check:
        if args.files:
            # No matching files in pre-commit mode
            return 0
        return 0

    # Collect violations
    violations: list[tuple[str, int]] = []

    for py_file in files_to_check:
        code_lines = count_code_lines(py_file)
        if code_lines > max_lines:
            rel_path = py_file.relative_to(root_dir)
            violations.append((str(rel_path), code_lines))

    # Output results
    if violations:
        print(f"[ERROR] The following files exceed {max_lines} code lines:")
        for path, lines in sorted(violations):
            print(f"  {path} -> {lines} lines")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
